### PR TITLE
Document NFD for GPU Labeling

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
This  patch documents the official k8s-sig NFD project for automated node labeling, and restructures the ` Automatic node labelling` section at `scheduling-gpus` page

### CODE REVIEW

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (mikemckiernan) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:72]: L#71 doesn't have as much info payload as #72, and I think we can combine:

"NFD detects the hardware features that are available on each node in a Kubernetes cluster and advertises those features. Typically, NFD adds node labels to advertise the features, but NFD can also add extended resources, annotations, and node taints."

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (mikemckiernan) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:73]: I'll rarely critique using a full project name, but here we've already abbreviated, so I start with "NFD is..."  However, I'd wait to hear from a maintainer about whether and how to mention version information.  The [Versioning Kubernetes examples](https://kubernetes.io/docs/contribute/style/style-guide/#versioning-kubernetes-examples) topic doesn't apply, I don't think.  v1.21 is already EOL, IIUC.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (mikemckiernan) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:67]: I think we'd want to keep, but modify this heading.  Maybe "Manage clusters with different types of GPUs".  My logic is that this is in the task section and starting headings with a verb, when possible, is best.

I'd also keep the rather generic info from the original L#69-71 because it mentions a simple mechanism for managing the diff devices--labels.  Then, under this heading, identify the original content as "manual":

> ## Manage clusters with different types of GPUs
> 
> If different nodes in your cluster have different types of GPUs, then you
> can use [Node Labels and Node Selectors](/docs/tasks/configure-pod-container/assign-pods-nodes/)
> to schedule pods to appropriate nodes.
> 
> ### Manually labeling nodes
> 
> You can manually apply labels to nodes and then reference the labels in your pod specification to place workloads on the nodes with the specific type of GPU.
> 
> ```console
> # Label your nodes with the accelerator type they have.
> kubectl label nodes node1 accelerator=example-gpu-x100
> kubectl label nodes node2 accelerator=other-gpu-k915
> ```
> 
> That label key `accelerator` is just an example.
> You can use a different label key.

The proposed "Automatically labeling nodes..." heading would need another #, I think.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (mikemckiernan) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:75]: Hrm, my instinct is to remove this sentence.  I liked that the page was basic--just labels.  But, I don't feel strongly about it and defer to SMEs and the doc maintainers.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (mikemckiernan) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:76]: nit: s/Once/After/ on the principle of "one word, one meaning."  I'd also spell out "specification" but I'm not intimate with the style guidelines.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (mikemckiernan) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:92]: nitiest nit: how about `gpu-vendor.example/gpu-product: "example-gpu-x100"`?  I feel like a greater difference to L#90 could be helpful.  (Though, I concede that this is an area that is unlikely to cause any confusion. Sigh.)

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (mikemckiernan) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:96]: wordsmithing: "NVIDIA implemented..." most conjugations of "to be" can be omitted.

**Review by mikemckiernan — COMMENTED** [review]
Nits and a (likely unwelcome) widening of the scope to identify some of the existing content as the manual method of labeling.  I might be waltzing you into work that the doc maintainers don't want, so let's see what they say.  Thanks as always for updating the docs!

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:109]: I prefer to keep it as is, since most of the time the deviceplugin and node label match to make things simpler

**Comment by sftim** [comment]
Thanks

/sig node

**Review by sftim — COMMENTED** [review]
Hi; thanks for the pull request.

Here's my feedback on the changes.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (eero-t) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:92]: IMHO it's relevant enough to keep here, and adding link to docs is good also.

PS. NFD taint support is still experimental: https://nfd.sigs.k8s.io/usage/customization-guide#node-tainting

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (eero-t) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:120]: NFD alone is not enough, one also needs a config with suitable rules.  After quick googling, all 3 vendors' GPU operators would already seem to support NFD at least as optional dependency => I think removing this section is OK.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (eero-t) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:89]: Maybe:
```suggestion
NFD detects the hardware features that are available on each node in a Kubernetes cluster.
Typically, NFD is configured to advertise those features as node labels, but NFD can also add extended resources, annotations, and node taints.
```

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (eero-t) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:133]: List both AMD & NVIDIA implementations, or neither, don't remove just one.

(Intel has used NFD from start, so it's not relevant here.)

Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:133]: AMD doesn't implement NFD, they have a custom `node-labeller` that doesn't leverage NFD api.

Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:133]: https://github.com/ROCm/k8s-device-plugin/tree/master/cmd/k8s-node-labeller , is a sub component of their device plugin implementation

Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:133]: Should it still be included then?

Comment (eero-t) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:133]: > AMD doesn't implement NFD, they have a custom `node-labeller` that doesn't leverage NFD api.

How such implementation detail is relevant to users i.e. which custom labeling solutions this section should list?

Comment (sftim) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:133]: > List both AMD & NVIDIA implementations, or neither, don't remove just one.

That's not Kubernetes' policy; please disregard this suggestion @ArangoGutierrez.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (sftim) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:117]:
```suggestion
            values:
            - Helium # example product family
            - Neon   # example product family
```

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (sftim) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:128]:
```suggestion
```

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:133]: So only list projects that leverage NFD API?

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (sftim) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:134]:
```suggestion
```

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (sftim) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:132]: Is there an Intel implementation? If so, we should add a link (the entries are sorted alphabetically).

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (sftim) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:131]: Yep, exactly that.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:130]: K8s-sig/NFD main contributors are INTEL and NVIDIA. For very specific vendored labels NVIDIA uses the gpu-feature-discovery sidecar, which calls to NFD API to generate `nvidia.com/` labels.
For all INTEL features, NFD generates them with `feature.node.kubernetes.io/cpu-cpuid.ADX` as example.

Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:126]: is there a GFD version at INTEL @marquiz  ??

Comment (marquiz) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:126]: It's the Intel device plugins for kubernetes (https://github.com/intel/intel-device-plugins-for-kubernetes) that's I think already linked above. Anything to add @eero-t ?

Comment (sftim) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:126]: Please list it here too, if it works as a plugin to NFD.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (stmcginnis) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:128]: nit: NVIDIA's actual name is in all caps, but Intel is normal cased, so it would be good to just have it be Intel and not shout INTEL. :)

Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:128]: My bad, use of habit

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (eero-t) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:126]: I'm not sure what the above mentioned "NFD API" or "plugin to NFD" (vendor specific fork of NFD worker?) mean, as NFD is configured through NFD label/etc rules and runtime-provided feature (config) file(s) provided to the per-node NFD workers.

Intel uses following:
* Intel device plugins project has NFD config rules for device plugin DaemonSets: https://github.com/intel/intel-device-plugins-for-kubernetes/tree/main/deployments/nfd/overlays/node-feature-rules/
* Deployed GPU plugin will then provide NFD feature file for additional, more targeted labels [1]: https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/cmd/gpu_plugin/labels.md

[1] These labels were earlier provided by (deprecated) GPU plugin init container NFD hook binary [2], but hooking support will be removed in next NFD version [3].

[2] https://intel.github.io/intel-device-plugins-for-kubernetes/cmd/gpu_nfdhook/README.html

[3] https://github.com/kubernetes-sigs/node-feature-discovery/issues/856

**Comment by sftim** [comment]
/hold

I have **unresolved** one outstanding bit of feedback:
- https://github.com/kubernetes/website/pull/44915/files#r1468853640

Please address this. OK to unhold once that's addressed. Please avoid marking it resolved without responding to the feedback.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (sftim) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:128]: Alphabetical order please
```suggestion

These vendors provide labelling implementations that integrate with NFD for GPU hardware discovery:

- [Intel](https://github.com/intel/intel-device-plugins-for-kubernetes).
- [NVIDIA](https://github.com/NVIDIA/gpu-feature-discovery/#readme).
```

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:96]: You mean adding something like ?

```
#
# CAUTION: this is an example configuration.
#          Do not use this for your own cluster!
#
```

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (eero-t) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:128]: As this is only about GPUs, it's better to refer directly to Intel GPU plugin documentation: https://intel.github.io/intel-device-plugins-for-kubernetes/cmd/gpu_plugin/README.html

**Review by eero-t — APPROVED** [review]

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (stmcginnis) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:100]: I believe what @sftim was pointing out is there is a "shortcode" that is used to highlight significant parts of a code block.

So if you look at [the raw markdown](https://raw.githubusercontent.com/kubernetes/website/main/content/en/docs/tasks/administer-cluster/encrypt-data.md) for the references above, you can see there is this line:

```txt
{{< highlight yaml "linenos=false,hl_lines=2-5" >}}
```

That is telling the rendering to not include line numbers (not as important) and to highlight lines 2 through 5 in the yaml content. So you could use that to highlight the affinity section that is relevant for selecting the right hosts based on the desired GPU memory and family.

Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:100]: Oh I see

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (sftim) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:99]:
```suggestion
```

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (sftim) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:106]:
```suggestion
  # You can use Kubernetes node affinity to schedule this Pod onto a node
  # that provides the kind of GPU that its container needs in order to work
```

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:96]: I think is done now

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (stmcginnis) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:95]:
```suggestion
{{< highlight yaml "linenos=false,hl_lines=6-20" >}}
```

**Comment by sftim** [comment]
Thanks @ArangoGutierrez 

/hold cancel

This is ready for technical review.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (stmcginnis) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:127]: nit - only if you need to do another update for a more compelling reason: I would remove the period from the end of these bullet items since it is just a list of the names/hyperlinks.

**Review by stmcginnis — COMMENTED** [review]
/lgtm

Preview:
https://deploy-preview-44915--kubernetes-io-main-staging.netlify.app/docs/tasks/manage-gpus/scheduling-gpus/

Looks great, thanks for working through all the feedback.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md:127]: I always had the question of wetter to end or not with a `.` on MD lists 
Thanks!

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (klueska) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:123]: This seems to just be an extension of the example above, with the added affinity section, correct?

However, from reading the docs top-to-bottom, this implies that a label called `gpu.gpu-vendor.example/installed-memory`will somehow magically be added to the node by NFD (which isn't true, an NFD plugin is needed to add this).

We should extend the text to make it clear that an NFD plugin is needed (in addition to NFD itself) in order to provide any vendor specific labels.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (sftim) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:116]:
```suggestion
  restartPolicy: OnFailure
```
to match the earlier example.

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (sftim) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:91]:
```suggestion
You also need a plugin for NFD that adds labels to your nodes. Your GPU vendor may provide a third party
plugin for NFD; check their documentation for more details.

```

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md]: How about now?

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (ArangoGutierrez) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md]: I did some mods to this suggestion please check

Comment (eero-t) [reply content/en/docs/tasks/manage-gpus/scheduling-gpus.md]: I did not find any mentions of "NFD plugin" from NFD documentation:  https://kubernetes-sigs.github.io/node-feature-discovery

Whereas NFD release notes mentioned "NFD plugin" to be functionality that NFD provides for `kubelet`.

If "NFD plugin" in above means those "hook" binaries that NFD worker called earlier, that support is disabled by default, and will be removed in next NFD release (due to security reasons), so I think some better term could be used.

Something like this sounds a bit too much though:
> NFD alone does not label nodes, example here requires additional vendor support, either NFD configuration specifying device matching rules it will use for labeling, or additional component directly providing NFD node worker list of labels.

**Review by sftim — COMMENTED** [review]
This looks right.

- https://github.com/kubernetes/website/pull/44915#pullrequestreview-1851478814 added LGTM
- https://github.com/kubernetes/website/pull/44915#discussion_r1471508671 provided feedback
- I judge that the feedback is addressed

It looks OK to publish this as a change to the live site.

/lgtm
/approve

Thanks everyone!

File: content/en/docs/tasks/manage-gpus/scheduling-gpus.md
Comment (sftim) [inline content/en/docs/tasks/manage-gpus/scheduling-gpus.md:98]:
```suggestion
{{< highlight yaml "linenos=false,hl_lines=6-18" >}}
```

**Comment by sftim** [comment]
Actually I fixed some highlighting.

**Comment by ArangoGutierrez** [comment]
> OK, it's less wrong now. /lgtm

Thanks a lot!

**Pull request merged**
